devito/passes/clusters/aliases.py

```diff
@@ -485,91 +543,89 @@ def collect(extracted, ispace, min_storage):
 
-    # Do not waste time if unneccesary
-    if not candidates:
-        return exprs, retained, tot
+    # Filter off the aliases with low score
+    key = lambda a: a.score >= m
+    aliases.filter(key)
 
-    # Project the candidate aliases into exprs to determine what the new
-    # working set would be
-    mapper = {k: v for k, v in mapper.items() if v.free_symbols & set(aliaseds)}
+    # Project the candidate aliases into `exprs` to derive the final working set
+    mapper = {k: v for k, v in mapper.items() if v.free_symbols & set(aliases.aliaseds)}
     templated = [uxreplace(e, mapper) for e in exprs]
+    owset = wset(templated)
 
-    # Pass 2: a set of aliasing expressions is retained only if the tradeoff
-    # between operation count reduction and working set increase is favorable
-    owset = wset(others + templated)
-    for e, v in aliases.items():
-        try:
-            score = candidates[e]
-        except KeyError:
-            score = 0
-        if score > 1 or \
-           score == 1 and max(len(wset(e)), 1) > len(wset(e) & owset):
-            retained[e] = v
-            tot += score
+    # Filter off the aliases with a weak flop-reduction / working-set tradeoff
+    key = lambda a: \
+        a.score > M or \
+        m <= a.score <= M and max(len(wset(a.pivot)), 1) > len(wset(a.pivot) & owset)
+    aliases.filter(key)
 
-    # Do not waste time if unneccesary
-    if not retained:
-        return exprs, retained, tot
+    if not aliases:
+        return exprs, aliases
 
     # Substitute the chosen aliasing sub-expressions
-    mapper = {k: v for k, v in mapper.items() if v.free_symbols & set(retained.aliaseds)}
+    mapper = {k: v for k, v in mapper.items() if v.free_symbols & set(aliases.aliaseds)}
     exprs = [uxreplace(e, mapper) for e in exprs]
 
-    return exprs, retained, tot
+    return exprs, aliases
 
 
-def lower_aliases(cluster, aliases, in_writeto, maxpar):
+def lower_aliases(aliases, meta, maxpar):
     """
-    Create a Schedule from an AliasMapper.
+    Create a Schedule from an AliasList.
     """
     dmapper = {}
     processed = []
-    for alias, v in aliases.items():
-        imapper = {**{i.dim: i for i in v.intervals},
-                   **{i.dim.parent: i for i in v.intervals if i.dim.is_NonlinearDerived}}
+    for a in aliases:
+        imapper = {**{i.dim: i for i in a.intervals},
+                   **{i.dim.parent: i for i in a.intervals if i.dim.is_NonlinearDerived}}
 
         intervals = []
         writeto = []
         sub_iterators = {}
-        indicess = [[] for _ in v.distances]
-        for i in cluster.ispace.intervals:
+        indicess = [[] for _ in a.distances]
+        for i in meta.ispace:
             try:
                 interval = imapper[i.dim]
             except KeyError:
-                # E.g., `x0_blk0` or (`a[y_m+1]` => `y not in imapper`)
-                intervals.append(i)
-                continue
+                if i.dim in a.free_symbols:
```

Comment:
Note for reviewers: this is a small bug fix

Reply:
What does it fix? What is the bug?

Reply:
I don't remember exactly anymore, but I've added a test. It had to do with hierarchical blocking and maybe nested parallelism